File: src/com/mantono/webserver/reflection/ResourceFinder.java
package com.mantono.webserver.reflection;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class ResourceFinder
{
	private final static String CLASS = "\\w+\\.class\\b";
	private final String[] classPath;
	private List<Path> pathResourceClasses;
	private List<Class<?>> classes = new ArrayList<Class<?>>();

	public ResourceFinder()
	{
		classPath = System.getProperty("java.class.path").split(":");
	}

	public int search() throws IOException
	{
		pathResourceClasses = new ArrayList<Path>();
		for(String path : classPath)
		{
			Stream<Path> paths = Files.walk(createPath(path));
			List<Path> foundResources = paths
				.filter(t -> isClassFile(t.getFileName()))
				.filter(c -> isResource(c))
				.collect(Collectors.toList());
			pathResourceClasses.addAll(foundResources);
			paths.close();
		}

		return pathResourceClasses.size();
	}

	private Path createPath(String path)
	{
		return new File(path).toPath();
	}

	private boolean isResource(Path c)
	{
		try
		{
			String className = getQualifiedClassName(c);
			Class<?> loadedClass = Class.forName(className);
			return ClassParser.hasResources(loadedClass);
		}
		catch(ClassNotFoundException e)
		{
			e.printStackTrace();
		}
		return false;
	}

	private String getQualifiedClassName(Path c)
	{
		String className = removeClassPath(c.toString());
		className = className.substring(0, className.length()-6);
		className = className.replaceAll("\\/", ".");
		className = className.replaceFirst("\\B\\.", "");
		return className;
	}

	private String removeClassPath(String resource)
	{
		for(String path : classPath)
			if(resource.contains(path))
				return resource.replaceFirst(path, "");

		return null;
	}

	private boolean isClassFile(Path fileName)
	{
		Pattern p = Pattern.compile(CLASS);
		Matcher m = p.matcher(fileName.toString());
		return m.find();
	}

	private void loadClasses() throws ClassNotFoundException
	{
		for(Path path : pathResourceClasses)
		{
			Class<?> loadedClass = Class.forName(getQualifiedClassName(path));
			classes.add(loadedClass);
		}
	}

	public List<Class<?>> getClasses() throws ClassNotFoundException
	{
		loadClasses();
		return classes;
	}
}

File: src/main/java/com/mantono/webserver/reflection/resourceFinder.kt
package com.mantono.webserver.reflection

import com.mantono.webserver.ValidRequest
import com.mantono.webserver.rest.ContentType
import com.mantono.webserver.rest.Resource
import com.mantono.webserver.rest.Verb
import sun.misc.Request
import java.io.File
import java.lang.reflect.Method
import java.nio.file.Files
import java.nio.file.Path
import java.util.Arrays.stream
import java.util.stream.Stream
import kotlin.streams.asSequence
import kotlin.streams.toList

private val CLASS: Regex = Regex("\\w+\\.class\\b")
val classPath: List<String> = System.getProperty("java.class.path").split(":".toRegex())

val staticResources: Map<String, Path> by lazy {

	val resourceRegex: Regex = Regex("/resources(/|$)")
	val resourceDirs: List<String> = classPath.filter { resourceRegex.containsMatchIn(it) }

	resourceDirs
			.map { File(it).toPath() }
			.map { Files.walk(it)}
			.flatMap { it.toList() }
			.map { withoutPath(it, resourceDirs) }
			.filter { it.first.isNotEmpty() }
			.toMap()
}

fun withoutPath(entirePath: Path, resourceDirs: List<String>): Pair<String, Path>
{
	val root: String = resourceDirs.firstOrNull{ it in entirePath.toString() }!!
	val subPath: String = entirePath.toString().removePrefix(root)
	return Pair(subPath, entirePath)
}

fun findResources(path: List<String> = classPath): Map<Resource, Method>
{
	return  methodsByAnnotation(Resource::class.java, path)
			.filter { hasCorrectParameters(it) }
			.map { pairOfResourceMethod(it) }
			.asSequence()
			.toMap()
}

fun <T : Annotation> methodsByAnnotation(annotation: Class<T>, path: List<String> = classPath): Stream<Method>
{
	return classesInPath(path)
			.map { it.declaredMethods }
			.flatMap(::stream)
			.filter { it.isAnnotationPresent(annotation) }
}

fun classesInPath(path: List<String> = classPath): Stream<Class<*>>
{
	return  path.stream()
			.parallel()
			.map { File(it).toPath() }
			.filter { Files.exists(it) }
			.map { Files.walk(it) }
			.flatMap { it }
			.filter { isClassFile(it.fileName)}
			.map { asClass(it) }
			.filterNull()
}

fun pairOfResourceMethod(method: Method): Pair<Resource, Method>
{
	val resource: Resource = method.getAnnotationsByType(Resource::class.java)[0]
	return Pair(resource, method)
}

fun hasCorrectParameters(method: Method): Boolean
{
	val types: Array<out Class<*>> = method.parameterTypes
	val count: Int = method.parameterCount
	return when(count)
	{
		0 -> true
		1 -> return types[0] == Request::class.java || types[0] == ValidRequest::class.java
		else -> false
	}
}

private fun <T> Stream<T?>.filterNull(): Stream<T> = filter { it != null }.map { it!! }
private fun isClassFile(fileName: Path): Boolean = CLASS.containsMatchIn(fileName.toString())

fun asClass(pathClass: Path): Class<*>?
{
	val className = getQualifiedClassName(pathClass)
	return Class.forName(className)
}

private fun getQualifiedClassName(c: Path): String
{
	var className: String = removeClassPath(c.toString()) ?: c.toString()
	className = className.substring(0, className.length - 6)
	className = className.replace("\\/".toRegex(), ".")
	className = className.replaceFirst("\\B\\.".toRegex(), "")
	return className
}

private fun removeClassPath(resource: String): String?
{
	return classPath.firstOrNull {resource.contains(it)}
			?.let {resource.replaceFirst(it.toRegex(), "")}
}